sql: prohibit DROP COLUMN on REGIONAL BY ROW column reference

otan opened the pull request:
The foundation of REGIONAL BY ROW is to use a column which dictates the
location to which the row is stored. As such, we should prevent the
column from being used as a DROP COLUMN reference.

Resolves #59117 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60600)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
Ultra nit: In the commit record, it might be more correct to say "the location _in_ which the rows is stored".

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60600#-:-MTepKtGEJaiYC8Q2FKZ:bnxe66x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/alter_table.go, line 409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60600#-MTeqcOg1262N5-ajVPx:-MTeqcOg1262N5-ajVPy:b-pdp0rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a5de651873d9a0653b117a4144f1ba8b1bb1785/pkg/sql/alter_table.go#L409)):*
> ```Go
> 						pgerror.Newf(
> 							pgcode.InvalidColumnReference,
> 							"cannot drop column %s as it is used as the REGIONAL BY ROW column reference",
> ```

Nit: "as it is used to store the region in a REGIONAL BY ROW table"?
___
*[pkg/sql/alter_table.go, line 412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60600#-MTequeE5fAO2aLDo2sI:-MTequeE5fAO2aLDo2sJ:bysyohj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a5de651873d9a0653b117a4144f1ba8b1bb1785/pkg/sql/alter_table.go#L412)):*
> ```Go
> 							t.Column,
> 						),
> 						"You must change the table locality before dropping this table.",
> ```

Nit: Or alter the table to choose a different column to use for the region info?  
___
*[pkg/sql/catalog/tabledesc/structured.go, line 4029 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60600#-MTerU3s93h7N_2jhNk4:-MTerU3s93h7N_2jhNk5:b-3702f9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a5de651873d9a0653b117a4144f1ba8b1bb1785/pkg/sql/catalog/tabledesc/structured.go#L4029)):*
> ```Go
> func (desc *wrapper) GetRegionalByRowTableRegionColumnName() (tree.Name, error) {
> 	if !desc.IsLocalityRegionalByRow() {
> 		return "", errors.Newf("%s is not REGIONAL BY ROW", desc.Name)
> ```

I'd think that this should be an Assertf.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm

otan commented:
bors r-

otan committed:
sql: prohibit DROP COLUMN on REGIONAL BY ROW column reference

The foundation of REGIONAL BY ROW is to use a column which dictates the
location in which the row is stored. As such, we should prevent the
column from being used as a DROP COLUMN reference.

Release note: None

otan commented:
bors r=ajstorm
